File: android/app/src/main/java/com/dungeonboss/game/PartyCrawlResolver.kt
package com.dungeonboss.game

import com.dungeonboss.model.Encounter
import com.dungeonboss.model.Hero
import com.dungeonboss.model.PlacedRoom

/**
 * Runs a whole party through a dungeon, applying room effects. At each room, in
 * order: a poison tick, a one-shot delayed hit (Poisoned Spikes), the room's
 * party-wide hits (Antimagic/Zealots, unreducible), then a single-target hit on
 * the highest-health hero (reducible unless the room or a modifier says
 * otherwise). Single-target damage includes the room's grow bonus, dungeon auras
 * (Trap Makers / Beast Tamer), the boss's points bonus, and per-crawl modifiers.
 * A hero hit by a Poison Gas room is poisoned (+1 unreducible per later room). A
 * grow-on-death room gains +1 damage permanently per death there. A Retreat
 * modifier turns the party back before its target room. Stateless entry point;
 * uses an instance to carry crawl state. Mirrors
 * `webapp/lib/party_crawl_resolver.rb`.
 */
object PartyCrawlResolver {
    data class Step(
        val encounter: Encounter,
        val hero: Hero,
        val damage: Int,
        val healthAfter: Int,
        val died: Boolean,
        val roomIndex: Int
    )

    data class Result(
        val participants: List<Hero>,
        val log: List<Step>,
        val deaths: Int,
        val deadHeroes: List<Hero>,
        val survivors: List<Hero>,
        /** {"ability"|"room" -> count} the owner draws from death-triggered rooms. */
        val draws: Map<String, Int>
    )

    fun resolve(
        party: Party,
        dungeon: Dungeon,
        bossBonus: Int = 0,
        modifiers: CrawlModifiers = CrawlModifiers(),
        dryRun: Boolean = false
    ): Result = Run(party, dungeon, bossBonus, modifiers, dryRun).resolve()

    private class Run(
        party: Party,
        private val dungeon: Dungeon,
        private val bossBonus: Int,
        private val mods: CrawlModifiers,
        // A dry run predicts the outcome without the one lasting side effect —
        // grow-on-death rooms are not permanently grown — so it is safe to run
        // repeatedly (e.g. to preview a crawl before it is committed).
        private val dryRun: Boolean = false
    ) {
        private val participants: List<Hero> = party.heroes.toList()
        private val health = HashMap<Hero, Int>()
        private val alive = participants.toMutableList()
        private val dead = mutableListOf<Hero>()
        private val poison = HashMap<Hero, Int>()    // hero -> poison stacks
        private val delayed = HashMap<Hero, Int>()   // hero -> unreducible damage next room
        private val draws = HashMap<String, Int>()   // deck name -> cards owner draws
        private val log = mutableListOf<Step>()

        private var roomIndex = 0
        private var deathsHere = 0

        init {
            for (hero in participants) health[hero] = hero.maxHp
        }

        fun resolve(): Result {
            val encounters = dungeon.encounters()
            for ((index, encounter) in encounters.withIndex()) {
                if (alive.isEmpty()) break
                if (mods.retreatsAt(index)) break // Retreat: the party turns back here

                roomIndex = index
                deathsHere = 0

                poisonTick(encounter)
                if (alive.isNotEmpty()) delayedTick(encounter)
                if (alive.isNotEmpty()) applyPartyHits(encounter)
                if (alive.isNotEmpty()) applySingleTarget(encounter)
                grow(encounter)
                recordDeathDraws(encounter)
            }
            return Result(
                participants = participants,
                log = log,
                deaths = dead.size,
                deadHeroes = dead,
                survivors = alive,
                draws = draws
            )
        }

        /** Most current health, then most max health, then earliest board order. */
        private fun pickTarget(): Hero =
            alive.maxWith(compareBy({ health.getValue(it) }, { it.maxHp }))

        private fun poisonTick(encounter: Encounter) {
            for (hero in alive.toList()) {
                val stacks = poison[hero] ?: 0
                if (stacks > 0) hit(hero, stacks, encounter, reducible = false)
            }
        }

        private fun delayedTick(encounter: Encounter) {
            for (hero in alive.toList()) {
                val amount = delayed[hero] ?: 0
                if (amount > 0) {
                    hit(hero, amount, encounter, reducible = false)
                    delayed.remove(hero)
                }
            }
        }

        private fun applyPartyHits(encounter: Encounter) {
            RoomEffect.forEncounter(encounter).partyHits(alive.toList()).forEach { (hero, amount) ->
                if (alive.contains(hero)) hit(hero, amount, encounter, reducible = false)
            }
        }

        private fun applySingleTarget(encounter: Encounter) {
            val base = effectiveBase(encounter)
            if (base <= 0) return

            val effect = RoomEffect.forEncounter(encounter)
            // A per-crawl unreducible (Expose Weakness / Troll) OR the room's own
            // unreducible effect (Champion's Arena) makes this hit unreducible.
            val reducible = mods.reducible(roomIndex) && !effect.unreducible()

            if (effect.damagesAll()) {
                // The base hit lands on every alive hero (Poison Gas). Each hit is
                // reduced independently (so a Rogue protects the whole party), and
                // only heroes actually damaged are poisoned. Snapshot the list so
                // a hero killed by this room is not hit twice.
                for (hero in alive.toList()) {
                    applyHitWithEffects(hero, base, encounter, reducible, effect)
                }
                return
            }

            // Single-target hit on the highest-health hero. When the target dies,
            // the leftover (overkill) damage spills onto the next-highest hero, so
            // one big room can wipe a whole party.
            var remaining = base
            while (remaining > 0 && alive.isNotEmpty()) {
                val target = pickTarget()
                val before = health.getValue(target)
                val dealt = applyHitWithEffects(target, remaining, encounter, reducible, effect)
                if (dealt <= 0) break
                remaining = if (dealt > before) dealt - before else 0
            }
        }

        /** Hit one hero and apply this room's on-hit effects; returns damage dealt. */
        private fun applyHitWithEffects(
            hero: Hero,
            amount: Int,
            encounter: Encounter,
            reducible: Boolean,
            effect: RoomEffect.Spec
        ): Int {
            val dealt = hit(hero, amount, encounter, reducible)
            if (dealt <= 0) return 0
            if (alive.contains(hero)) {
                if (effect.poisonsOnHit()) poison[hero] = (poison[hero] ?: 0) + 1
                if (effect.nextRoomDamage() > 0) {
                    delayed[hero] = (delayed[hero] ?: 0) + effect.nextRoomDamage()
                }
            }
            return dealt
        }

        private fun recordDeathDraws(encounter: Encounter) {
            val deck = RoomEffect.forEncounter(encounter).drawsOnDeath()
            if (deck != null && deathsHere > 0) draws[deck] = (draws[deck] ?: 0) + deathsHere
        }

        private fun effectiveBase(encounter: Encounter): Int {
            if (mods.isZero(roomIndex)) return 0

            var base = if (mods.isSet(roomIndex)) mods.setValue(roomIndex) else encounter.damage
            val bossEffect = BossEffect.forBoss(dungeon.boss)
            base += if (encounter === dungeon.boss) {
                bossEffect.selfBonus(bossBonus)
            } else {
                bossEffect.roomBonus(encounter, bossBonus)
            }
            base += auraBonus(encounter)
            return base + mods.bonus(roomIndex)
        }

        /** Damage other rooms' auras grant to this room (never the granter itself). */
        private fun auraBonus(encounter: Encounter): Int =
            dungeon.rooms.sumOf { room ->
                if (room === encounter) 0 else RoomEffect.forEncounter(room).auraBonus(encounter, bossBonus)
            }

        private fun grow(encounter: Encounter) {
            if (dryRun) return // a preview must not permanently grow the room
            if (deathsHere <= 0) return
            if (!RoomEffect.forEncounter(encounter).growsOnDeath()) return
            if (encounter is PlacedRoom) encounter.grow += deathsHere
        }

        /** Deal damage to a hero and record it; returns the actual damage dealt. */
        private fun hit(hero: Hero, amount: Int, encounter: Encounter, reducible: Boolean): Int {
            if (amount <= 0) return 0

            val damage = if (reducible) {
                HeroAbility.damageTaken(hero, encounter, alive, amount)
            } else {
                amount
            }
            val after = health.getValue(hero) - damage
            health[hero] = after
            val died = after <= 0
            log.add(Step(encounter, hero, damage, after, died, roomIndex))
            if (died) {
                alive.removeAt(alive.indexOfFirst { it === hero })
                dead.add(hero)
                deathsHere += 1
            }
            return damage
        }
    }
}

File: webapp/lib/party_crawl_resolver.rb
# frozen_string_literal: true

require_relative "hero_ability"
require_relative "room_effect"
require_relative "boss_effect"
require_relative "crawl_modifiers"

# Runs a whole party through a dungeon, applying room effects. At each room, in
# order: a poison tick, the room's party-wide hits (Antimagic/Zealots,
# unreducible), then a single-target hit on the highest-health hero (reducible).
# Single-target damage includes the room's grow bonus, dungeon auras (Trap
# Makers / Hobgoblin Beastmaster), and the boss's points bonus. A hero damaged by a Poison
# Gas room is poisoned (+1 unreducible per later room). A hero that hits 0 dies;
# grow-on-death rooms gain +1 permanently per death there.
# Stateless entry point; uses an instance to carry crawl state.
class PartyCrawlResolver
  # participants: heroes who started, in board order
  # log: ordered [{ encounter:, hero:, damage:, health_after:, died:, room_index: }]
  # draws: {"ability"|"room" => count} the owner draws from death-triggered rooms
  Result = Struct.new(:participants, :log, :deaths, :dead_heroes, :survivors, :draws,
                      keyword_init: true)

  def self.resolve(party, dungeon, boss_bonus: 0, modifiers: CrawlModifiers.new)
    new(party, dungeon, boss_bonus, modifiers).resolve
  end

  def initialize(party, dungeon, boss_bonus, modifiers)
    @dungeon = dungeon
    @boss_bonus = boss_bonus
    @mods = modifiers
    @participants = party.heroes.dup
    @health = @participants.to_h { |hero| [hero, hero.health] }
    @alive = @participants.dup
    @dead = []
    @poison = Hash.new(0) # hero => poison stacks
    @delayed = Hash.new(0) # hero => unreducible damage to take in the NEXT room
    @draws = Hash.new(0)   # deck name ("ability"/"room") => cards the owner draws
    @log = []
  end

  def resolve
    @dungeon.encounters.each_with_index do |encounter, index|
      break if @alive.empty?
      break if @mods.retreats_at?(index) # Retreat: the party turns back here

      @room_index = index
      @deaths_here = 0

      poison_tick(encounter)
      delayed_tick(encounter) unless @alive.empty?
      apply_party_hits(encounter) unless @alive.empty?
      apply_single_target(encounter) unless @alive.empty?
      grow(encounter)
      record_death_draws(encounter)
    end

    Result.new(participants: @participants, log: @log, deaths: @dead.size,
               dead_heroes: @dead, survivors: @alive, draws: @draws)
  end

  private

  # Most current health, then most max health, then earliest board order.
  def pick_target
    @alive.max_by { |hero| [@health[hero], hero.health] }
  end

  def poison_tick(encounter)
    @alive.dup.each do |hero|
      stacks = @poison[hero]
      hit(hero, stacks, encounter, reducible: false) if stacks.positive?
    end
  end

  # One-shot delayed damage (e.g. Poisoned Spikes) lands as the hero enters the
  # next room, then clears.
  def delayed_tick(encounter)
    @alive.dup.each do |hero|
      amount = @delayed[hero]
      next unless amount.positive?

      hit(hero, amount, encounter, reducible: false)
      @delayed.delete(hero)
    end
  end

  def apply_party_hits(encounter)
    RoomEffect.for(encounter).party_hits(@alive.dup, encounter).each do |hero, amount|
      next unless @alive.include?(hero)

      hit(hero, amount, encounter, reducible: false)
    end
  end

  def apply_single_target(encounter)
    base = effective_base(encounter)
    return unless base.positive?

    effect = RoomEffect.for(encounter)
    # A per-crawl unreducible (Expose Weakness / Troll) OR the room's own
    # unreducible effect (Champion's Arena) makes this hit unreducible.
    reducible = @mods.reducible?(@room_index) && !effect.unreducible?
    target = pick_target
    dealt = hit(target, base, encounter, reducible: reducible)
    return unless dealt.positive?

    # A hero hit by a Poison Gas room is poisoned thereafter.
    @poison[target] += 1 if effect.poisons_on_hit?
    # A hero hit by a Poisoned Spikes room takes delayed damage in the next room.
    @delayed[target] += effect.next_room_damage if effect.next_room_damage.positive?
  end

  # When heroes die in a room with a draw-on-death effect, the owner draws.
  def record_death_draws(encounter)
    deck = RoomEffect.for(encounter).draws_on_death
    @draws[deck] += @deaths_here if deck && @deaths_here.positive?
  end

  def effective_base(encounter)
    return 0 if @mods.zero?(@room_index)

    base = @mods.set?(@room_index) ? @mods.set_value(@room_index) : encounter.damage
    boss_effect = BossEffect.for(@dungeon.boss)
    if encounter.equal?(@dungeon.boss)
      base += boss_effect.self_bonus(@boss_bonus) # boss's own per-point bonus
    else
      base += boss_effect.room_bonus(encounter, @boss_bonus) # boss aura on this room
    end
    base += aura_bonus(encounter)
    base + @mods.bonus(@room_index)
  end

  # Damage other rooms' auras grant to this room (never the granter itself).
  def aura_bonus(encounter)
    @dungeon.rooms.sum do |room|
      room.equal?(encounter) ? 0 : RoomEffect.for(room).aura_bonus(encounter, @boss_bonus)
    end
  end

  def grow(encounter)
    return unless RoomEffect.for(encounter).grows_on_death? && @deaths_here.positive?

    encounter.grow += @deaths_here
  end

  # Deal damage to a hero and record it; returns the actual damage dealt.
  def hit(hero, amount, encounter, reducible:)
    return 0 unless amount.positive?

    damage = reducible ? HeroAbility.damage_taken(hero, encounter, @alive, amount) : amount
    @health[hero] -= damage
    died = @health[hero] <= 0
    @log << { encounter: encounter, hero: hero, damage: damage,
              health_after: @health[hero], died: died, room_index: @room_index }
    if died
      @alive.delete(hero)
      @dead << hero
      @deaths_here += 1
    end
    damage
  end
end
